### PR TITLE
dpcpp ports amgx_pgm

### PR DESCRIPTION
**Pull request by yhmtsai**
This PR ports the amgx_pgm to dpcpp.
the current `thrust::tie` can not be translated to dpcpp by dpct.

@upsj Do you know something like `thrust::tie` inside oneDPL?

### CODE REVIEW

**Comment by upsj** [comment]
`std::tie` from `<tuple>`

**Review by Slaedr — APPROVED** [review]
LGTM! Looks completely consistent with the cuda implementation. Just one comment below.

**Review by upsj — APPROVED** [review]
LGTM as well! I agree with @Slaedr, and would even go one step further: All of these kernels use one thread per row. So you could simplify them heavily by putting them directly into a 1D parallel_for

File: dpcpp/multigrid/amgx_pgm_kernels.dp.cpp
Comment (upsj) [inline dpcpp/multigrid/amgx_pgm_kernels.dp.cpp:51]: seems unused?

**Comment by yhmtsai** [comment]
no, the extra is our own effort.
it gives the same structure as the cuda and hip call including the share memory usage.
when the kernel call many times, the function in dpcpp is hard to align with other executor

**Comment by yhmtsai** [comment]
@upsj @Slaedr I move those kernels to unified and add `device_std` as thrust for cuda/hip and as std for others

File: dpcpp/test/multigrid/amgx_pgm_kernels.cpp
Comment (Slaedr) [inline dpcpp/test/multigrid/amgx_pgm_kernels.cpp:62]: All these tests can now be move to `tests/multigrid`, I guess? Then you could remove `hip/test/multigrid/amgx_pgm_kernels.cpp` etc.

**Review by Slaedr — APPROVED** [review]
Excellent work!

Did you try benchmarking your previous implementations against these ones? Since we seem to be discussing multigrid performance quite often, it would be good to document the performance difference between the separate kernels and the unified ones. I guess it's unlikely there's a big difference, but it will be good to document it.

It might make sense to develop more fine-grain parallel kernels for the aggregation kernels, primarily to get good coalescing access, though I understand that is far from trivial. That can be left to another PR.

File: common/unified/multigrid/amgx_pgm_kernels.cpp
Comment (upsj) [reply common/unified/multigrid/amgx_pgm_kernels.cpp:146]: I tend to agree here - there is a clear path to using warp-parallelism here in the future. It is a cheap way to implement it in DPC++. There may be a few abstract primitives we could use for the GPU kernels: Csr row-wise reduction + counting, Csr row-wise filtering, ... Might make sense to put that into a file in `components`? But for now I wouldn't block the PR for this.
Probably should take a look at specializing it before the release?

Comment (yhmtsai) [reply common/unified/multigrid/amgx_pgm_kernels.cpp:146]: yes it should be parallelized by warp per row.
we do not yet test the performance of the generation step.
at least, the R*A*P also need to improve

Comment (Slaedr) [reply common/unified/multigrid/amgx_pgm_kernels.cpp:146]: My only concern is we might have to go back to backend-specific kernels if we need more advanced algorithms. But for now, I think we should move ahead with this.

**Comment by yhmtsai** [comment]
unified ~107.86ms
```
4.81%  3.1193ms         4  779.82us  129.02us  1.8408ms find_strongest_neighbor
0.05%  33.488us         4  8.3720us  6.4610us  10.141us match_edge
```

original ~107.108 (skip_sorting 62.4452)
```
4.82%  3.1674ms         4  791.86us  146.08us  1.8393ms find_strongest_neighbor
0.06%  36.687us         4  9.1710us  6.8120us  10.939us match_edge
```
only check mfem l-shape 1 level generation on V100
it does not give noticeable difference between these.
the main bottleneck is spgemm and sorting.
sorting takes ~ 40 ~ 45% of generation and spgemm takes ~10%
without sorting. spgemm (sum of RadixSort and spgemm_core) takes ~20% and find_strongest_neighbor takes ~10%

File: test/multigrid/amgx_pgm_kernels.cpp
Comment (upsj) [inline test/multigrid/amgx_pgm_kernels.cpp:173]: I guess this should be reverted?
```suggestion
    std::default_random_engine rand_engine;
```

Comment (yhmtsai) [reply test/multigrid/amgx_pgm_kernels.cpp]: good catch!

**Pull request merged**
